rust-src/concordium_base/src/web3id/mod.rs

```diff
@@ -0,0 +1,1540 @@
+pub mod did;
+
+// TODO:
+// - Documentation.
+use crate::{
+    base::CredentialRegistrationID,
+    common::base16_encode_string,
+    curve_arithmetic::Curve,
+    id::{
+        constants::{ArCurve, AttributeKind},
+        id_proof_types::{AtomicProof, AtomicStatement},
+        sigma_protocols::{self, vcom_eq::VecComEq},
+        types::{Attribute, AttributeTag, GlobalContext, IpIdentity},
+    },
+    pedersen_commitment::{self, VecCommitmentKey},
+    random_oracle::RandomOracle,
+};
+use concordium_contracts_common::{hashes::HashBytes, ContractAddress};
+use did::*;
+use ed25519_dalek::Verifier;
+use serde::de::DeserializeOwned;
+use std::collections::BTreeMap;
+use uuid::Uuid;
+
+/// A statement about a single credential, either an identity credential or a
+/// Web3 credential.
+#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq)]
+#[serde(
+    try_from = "serde_json::Value",
+    bound(deserialize = "C: Curve, AttributeType: Attribute<C::Scalar> + DeserializeOwned")
+)]
+pub enum CredentialStatement<C: Curve, AttributeType: Attribute<C::Scalar>> {
+    /// Statement about a credential derived from an identity issued by an
+    /// identity provider.
+    Identity {
+        network:   Network,
+        cred_id:   CredentialRegistrationID,
+        statement: Vec<AtomicStatement<C, u8, AttributeType>>,
+    },
+    /// Statement about a credential issued by a Web3 identity provider, a smart
+    /// contract.
+    Web3Id {
+        /// The credential type. This is chosen by the provider to provide
+        /// some information about what the credential is about. The list should
+        /// be considered as a "path", refining the meaning, e.g.,
+        /// "VerifiableCredential", "ConcordiumVerifiableCredential".
+        ty:         Vec<String>,
```

Comment:
It turns out that the standard actually insists on this vector as "an unordered set" https://www.w3.org/TR/vc-data-model/#dfn-type, but at the same time, types like `UniversityDegreeCredential` are considered "more narrow". I think it would make sense for type to be ordered and be some sort of subtyping relation.

Reply:
I don't know what exactly you are suggesting I should do here.

I can treat it as a set. Is that what needs to be done?

Reply:
I don't know what to do regarding how it's formulated in the standard either. I would say, that the idea of treating it as a "path" is better. I guess, we should not rely on the order in which these types are added to the list.

Reply:
This is now a set.